jormungandr/src/leadership/process.rs

```diff
@@ -0,0 +1,630 @@
+                        reason: "Failed to compute the schedule within time boundaries".to_owned()
+                    }).map_err(|()| unreachable!())
+                })
+                .map(|_| self);
+
+            Either::A(timeout)
+        } else {
+            // the event happened out of bounds, ignore it and move to the next one
+            error!(
+                logger,
+                "Eek... we missed an event schedule, system time might be off?"
+            );
+
+            let tell_user_about_failure = event_logs
+                .set_status(LeadershipLogStatus::Rejected {
+                    reason: "Missed the deadline to compute the schedule".to_owned(),
+                })
+                .map_err(|()| unreachable!());
+
+            Either::B(Either::B(tell_user_about_failure.map(|()| self)))
+        }
+    }
+
+    fn action_schedule(self) -> impl Future<Item = Self, Error = LeadershipError> {
+        let current_slot_position = self.current_slot_position().unwrap();
+
+        let epoch_tip = Epoch(self.tip_ref.block_date().epoch);
+
+        let logger = self.service_info.logger().new(o!(
+            "epoch_tip" => epoch_tip.0,
+            "current_epoch" => current_slot_position.epoch.0,
+            "current_slot" => current_slot_position.slot.0,
+        ));
+
+        if epoch_tip < current_slot_position.epoch {
+            let (leadership, _, _, _) =
+                new_epoch_leadership_from(current_slot_position.epoch.0, Arc::clone(&self.tip_ref));
+
+            let slot_start = current_slot_position.slot.0 + 1;
+            let nb_slots = leadership.era().slots_per_epoch() - slot_start;
+            let running_ref = leadership;
+
+            debug!(logger, "scheduling events" ;
+                "slot_start" => slot_start,
+                "nb_slots" => nb_slots,
+            );
+
+            Either::A(self.action_run_schedule(running_ref, slot_start, nb_slots))
+        } else if epoch_tip == current_slot_position.epoch {
+            // check for current epoch
+            let slot_start = current_slot_position.slot.0 + 1;
+            let nb_slots = self
+                .tip_ref
+                .epoch_leadership_schedule()
+                .era()
+                .slots_per_epoch()
+                - slot_start;
+            let running_ref = Arc::clone(self.tip_ref.epoch_leadership_schedule());
+
+            debug!(logger, "scheduling events" ;
+                "slot_start" => slot_start,
+                "nb_slots" => nb_slots,
+            );
+
+            Either::A(self.action_run_schedule(running_ref, slot_start, nb_slots))
+        } else {
+            // The only reason this would happen is if we had accepted a block
+            // that is set in the future or our system local date time is off
+
+            crit!(
+                logger,
+                "It seems the current epoch tip is way ahead of its time."
+            );
+            Either::B(future::ok(self))
```

Comment:
Is that really an `Ok`? If the node is supposed to go on after this, the log level should not be that severe.

Reply:
For now yes. It is critical for the stake pool operation. Not critical for the normal operation of the node.

Reply:
maybe an error then